### ISSUE TITLE
Compute variance of activations for masked tokens across diffusion timesteps

### ISSUE THREAD

**Issue by jds8** [jds8/SAE4DLM-CE#25]
For this task, we want to understand the spread of activation values at masked positions over the denoising process. The hypothesis is that activations early in the denoising process have higher variance (and lower entropy for the predicted unmasked token) than those later in the denoising process. Over 5 million text sequences, compute the [variance](https://docs.pytorch.org/docs/stable/generated/torch.var.html) of the activations at each denoising step, and generate a heatmap of features (y) vs denoising step (x). Let's do this for each of layers L1, L10, and L23 only for trainer 0 and only for masked tokens.

**Comment by jds8** [comment]
See the generate_history function here for how to extract the data from the DLM's history: https://github.com/jds8/SAE4DLM-CE/blob/LSOM/collect_activations.py#L343

**Comment by Dabsoysauce** [comment]
## Results: Variance of SAE Feature Activations at Masked Positions Across Diffusion Timesteps

### Setup
- **Model**: Dream-org/Dream-v0-Base-7B
- **SAEs**: `AwesomeInterpretability/dlm-mask-topk-sae` (top-k, 16,384 features)
- **Trainer**: trainer_0
- **Layers**: 1, 10, 23
- **Sequences**: 3,000 (FineWeb sample-10BT)
- **Timesteps**: 20 (mask probability p from 0.05 to 0.95)
- **Masked tokens only**
- **Variance computed via Welford's online algorithm**

### Summary

We computed the variance of SAE feature activations at masked token positions across 20 diffusion timesteps (mask probabilities from p=0.05 to p=0.95) for layers 1, 10, and 23 using trainer_0 SAEs on 3,000 text sequences from FineWeb. The results show a clear depth-dependent pattern: Layer 1 exhibits near-zero variance across all timesteps, indicating that early residual stream representations are largely insensitive to the noise level, while Layers 10 and 23 show a pronounced inverted-U relationship where mean feature variance rises from low masking probabilities, peaks around p≈0.57–0.62 (mid-denoising), and then declines again at high masking. Layer 23 consistently exhibits ~40% higher variance than Layer 10 at every timestep, with 15,658 active features compared to 11,296 and 10,656 for Layers 10 and 1 respectively. The per-feature z-normalized heatmaps reveal a striking diagonal structure across all layers: individual SAE features have distinct preferred noise levels where their variance peaks, and these preferences tile smoothly across the full range of timesteps, suggesting the SAE has learned features specialized to different stages of the denoising process. The variance distribution is highly skewed — at Layer 23, the top 3 features (F1311, F5949, F7116) account for variance values of 23,660, 12,500, and 1,478 respectively, while the median feature variance remains near zero across all timesteps, meaning the mean is driven almost entirely by a small number of high-variance outlier features. Contrary to the original hypothesis that activations early in denoising would have higher variance, the data instead suggests that mid-denoising is the point of maximum feature activation variance, which is intuitively consistent with the model facing the greatest uncertainty when it has partial but ambiguous context about the masked tokens.

### Key Plots

<img width="2000" height="1200" alt="Image" src="https://github.com/user-attachments/assets/932719ad-c5bb-4a8f-bdd7-c12262f7b7d4" />

<img width="2800" height="1600" alt="Image" src="https://github.com/user-attachments/assets/4721784c-2fb4-462d-b286-b363f15850de" />

<img width="2800" height="1600" alt="Image" src="https://github.com/user-attachments/assets/10705749-bb5f-4024-b6a7-af571a24a649" />

<img width="2800" height="1600" alt="Image" src="https://github.com/user-attachments/assets/f9542a8f-6fde-452b-9f6f-4c1614213952" />

<img width="2800" height="2000" alt="Image" src="https://github.com/user-attachments/assets/0ab67402-6e7e-4ce8-8f41-33a5e5dbb858" />

<img width="2000" height="1200" alt="Image" src="https://github.com/user-attachments/assets/ddb22d2e-6b5e-4cda-8fd0-f0e34fc8376f" />

**Cross-layer comparison:**
- `lineplot_cross_layer.png` — Inverted-U curve; L23 > L10 >> L1

**Per-feature z-normalized heatmaps (most informative):**
- `heatmap_L1_znorm.png`
- `heatmap_L10_znorm.png`
- `heatmap_L23_znorm.png`
- Diagonal structure shows features specialize to different denoising stages

**Top 50 highest-variance features:**
- `heatmap_L23_top50.png` — Top features peak at mid-denoising (p≈0.6)

**Variance distribution line plots:**
- `lineplot_L10.png` — Mean is driven by outliers; median near zero

**Comment by jds8** [comment]
These plots are really striking. The DLM-Scope SAEs are only trained on diffusion times from 0.05 up to 0.5, so I'm not sure what to make of the plots past t=0.5. Also, is there any reason you ran this on FineWeb rather than Common Pile? Could you create a branch (on the right of this page under Development) and merge your code into it so I can look into this more?
